### ISSUE TITLE
[feature] support creating a draft release

### ISSUE THREAD

**Issue by ianlewis** [slsa-framework/slsa-github-generator#1476]
The Go builder and generic generator use `softprops/action-gh-release` to create releases. We should support setting the `draft` flag so that users can create draft releases.

Related: https://github.com/sigstore/helm-sigstore/pull/111

**Comment by haydentherapper** [comment]
We ran into this in https://github.com/sigstore/timestamp-authority.

See the release GHA - https://github.com/sigstore/timestamp-authority/blob/main/.github/workflows/release.yaml

The GHA first creates a draft release with the binary assets, and then runs the provenance generator. The generator will create a release (not a draft) using the tag. The issues are:
1) This doesn't detect the existing draft release, instead creating a new release
2) This creates a new release that we have to delete after manually moving the intoto attestation to the draft release

Ideally the GHA generator would either detect an existing draft release and append the attestation or I could provide flags to the generator to control this behavior.

cc @asraa @laurentsimon

**Comment by asraa** [comment]
I *think* as a mitigation (suggested by Hayden!) users could use `upload-assets: false`, and manually upload the asset (which will be named according to the inputs: https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/README.md#workflow-inputs)

**Comment by ianlewis** [comment]
> I _think_ as a mitigation (suggested by Hayden!) users could use `upload-assets: false`, and manually upload the asset (which will be named according to the inputs: https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/README.md#workflow-inputs)

Yes, setting `upload-assets: false` and uploading them yourself (and downloading/uploading the provenance) is the workaround.
Example: https://github.com/sigstore/helm-sigstore/pull/111/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R61-R90

**Comment by haydentherapper** [comment]
Thanks, doing that here! https://github.com/sigstore/timestamp-authority/pull/215

**Comment by haydentherapper** [comment]
Thank you @ianlewis for the code pointer, a test release worked as expected!

**Comment by developer-guy** [comment]
While working on the osv-scanner project, I realized that they mark their release as a [draft](https://github.com/google/osv-scanner/blob/284a975e46e674c86547d809d7ecd5b486bfcd61/.goreleaser.yml#L86) for some reason. In that case, if we set the `upload-assets` parameter of the Generic SLSA 3 Generator as true, we see two different releases with the same tag, one is marked as draft the other is marked as latest. Then I dag into the problem a bit, and noticed that the Generic SLSA 3 Generator uses [soft-props/action-gh-release](https://github.com/softprops/action-gh-release) to upload the provenance file, [here](https://github.com/slsa-framework/slsa-github-generator/blob/66b541d35741996cc067fc2a56824c5a37f716ff/.github/workflows/generator_generic_slsa3.yml#L247) and [here](https://github.com/slsa-framework/slsa-github-generator/blob/66b541d35741996cc067fc2a56824c5a37f716ff/.github/workflows/generator_generic_slsa3.yml#L239). There is a draft option available in the [parameters](https://github.com/softprops/action-gh-release#-customizing) of the `action-gh-release` GitHub Action.

So, maybe we can add a new parameter to these generators for people who might want to mark their release as a draft, if so, they can pass that parameter to the `action-gh-release` to avoid having duplicate releases.

**Comment by developer-guy** [comment]
What I have thought is that;
* Add a `draft` option to the workflow inputs
* Use the `draft` option to fulfill the `draft` option of  [soft-props/action-gh-release](https://github.com/softprops/action-gh-release) that we used to upload provenance

does that sound good @ianlewis, if so, please assign it to me.

**Comment by ianlewis** [comment]
> What I have thought is that;
> 
> * Add a `draft` option to the workflow inputs
> * Use the `draft` option to fulfill the `draft` option of  [soft-props/action-gh-release](https://github.com/softprops/action-gh-release) that we used to upload provenance
> 
> does that sound good @ianlewis, if so, please assign it to me. PTAL @ianlewis

Yep, that's the gist of it I think.

**Comment by laurentsimon** [comment]
We need end-to-end test before closing this issue.

**Comment by ianlewis** [comment]
Reopen for e2e test.

**Comment by developer-guy** [comment]
Here is the PR for e2e tests: _https://github.com/slsa-framework/example-package/pull/167_

**Comment by SoulPancake** [comment]
Both these workarounds work btw
1. https://github.com/sigstore/timestamp-authority/pull/215/changes
2. As well as using the tag https://github.com/sigstore/timestamp-authority/pull/215/changes

Thanks everyone @ianlewis @Hayden-IO 
I was stalking a bunch of repos to eventually make it work.
This is useful when we need to maintain immutable releases/tags.

**Comment by SoulPancake** [comment]
However I feel we need to have an option to use existing draft release to update along with the draft option itself
Something like how goreleaser does it https://github.com/goreleaser/goreleaser/blob/49025ca384af2286db640fef2328c02d07c84ea8/www/content/customization/publish/scm/_index.md?plain=1#L45
If there is still interest I could open a PR to achieve this.
